akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala

```diff
@@ -0,0 +1,140 @@
+package akka.http.impl.engine.client
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.Done
+import akka.actor.{ Actor, ActorRef, DeadLetterSuppression, Deploy, NoSerializationVerificationNeeded, Props, Terminated }
+import akka.http.impl.engine.client.PoolInterfaceActor.PoolRequest
+import akka.http.impl.settings.HostConnectionPoolSetup
+import akka.http.scaladsl.HttpExt
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
+import akka.stream.Materializer
+
+import scala.concurrent.Promise
+
+/**
+ * Manages access to a host connection pool or rather: a sequence of pool incarnations.
+ *
+ * A host connection pool for a given [[HostConnectionPoolSetup]] is a running stream, whose outside interface is
+ * provided by its [[PoolInterfaceActor]] actor. The actor accepts [[PoolInterfaceActor.PoolRequest]] messages
+ * and completes their `responsePromise` whenever the respective response has been received (or an error occurred).
+ *
+ * The [[PoolMasterActor]] provides a layer of indirection between a [[PoolGateway]], which represents a pool,
+ * and the [[PoolInterfaceActor]] instances which are created on-demand and stopped after an idle-timeout.
+ *
+ * Several [[PoolGateway]] objects may be mapped to the same pool if they have the same [[HostConnectionPoolSetup]]
+ * and are marked as being shared. This is the case for example for gateways obtained through
+ * [[HttpExt.cachedHostConnectionPool]]. Some other gateways are not shared, such as those obtained through
+ * [[HttpExt.newHostConnectionPool]], and will have their dedicated restartable pool.
+ *
+ */
+private[http] class PoolMasterActor extends Actor {
+
+  import PoolMasterActor._
+
+  private[this] implicit val dispatcher = context.dispatcher
+
+  private[this] var poolStatus = Map[PoolGateway, PoolInterfaceStatus]()
+  private[this] var poolInterfaces = Map[ActorRef, PoolGateway]()
+
+  /**
+   * Start a new pool interface actor, register it in our maps, and watch its death.
+   *
+   * @param gateway the pool gateway this pool corresponds to
+   * @param fm the materializer to use for this pool
+   * @return the newly created actor ref
+   */
+  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): ActorRef = {
+    val props = Props(new PoolInterfaceActor(gateway)).withDeploy(Deploy.local)
+    val ref = context.system.actorOf(props, PoolInterfaceActor.name.next())
+    poolStatus += gateway -> PoolInterfaceRunning(ref)
+    poolInterfaces += ref -> gateway
+    context.watch(ref)
+  }
+
+  def receive = {
+
+    // Start or restart a pool without sending it a request. This is used to ensure that
+    // freshly created pools will be ready to serve requests immediately.
+    case s @ StartPool(gateway, materializer) ⇒
```

Comment:
It is not really one, and moreover it handles all the interface actors at once, so this would neither be easy nor elegant to model with a FSM. However, I will add comments describing the life cycle of a pool interface actor as seen by the pool master actor.